crates/sui-indexer/migrations_v2/2023-08-19-044026_transactions/up.sql

```diff
@@ -0,0 +1,20 @@
+CREATE TABLE transactions (
+    tx_sequence_number          BIGINT       PRIMARY KEY,
+    transaction_digest          bytea        NOT NULL,
```

Comment:
We noticed in #13718 that the JSON-RPC representation of a transaction (block) doesn't contain its effects digest. Currently, the GraphQL API currently exposes this, as well as a way to query transaction blocks by this digest:

https://github.com/MystenLabs/sui/blob/16730302d46d2836a8b8c523e350f98a78f6b5b4/crates/sui-graphql-rpc/schema/draft_schema.graphql#L73-L75

One reason for this is that it's a more direct transaction of the storage model (that effects are keyed by their own digest) and the other is that the execution API returns the effects digest:

https://github.com/MystenLabs/sui/blob/16730302d46d2836a8b8c523e350f98a78f6b5b4/crates/sui-graphql-rpc/schema/draft_schema.graphql#L149-L165

https://github.com/MystenLabs/sui/blob/16730302d46d2836a8b8c523e350f98a78f6b5b4/crates/sui-graphql-rpc/schema/draft_schema.graphql#L1026-L1029

The `executionBlock` mutation returns when the transaction is final, and the thinking is that if a client wants to confirm that the indexer has caught up, it can poll the indexer for the transaction block by its effects digest.  We could just use the transaction digest for this polling as well, but opted for the effects digest because it's more of a clue that the transaction has run.

In Indexer V2 + GraphQL we have **two options** then:

- Store the effects digest and an index of transactions by their effects digest, and retain the flow we currently designed for in the GraphQL API.
- Don't store the effects digest or corresponding index, remove the effects digest from `TransactionBlockEffects` and switch `executionBlock` to returning the transaction digest in the GraphQL API.

Let me know what you think!  I'm happy with either option, but have a slight preference for the current set-up on the GraphQL side.

Reply:
> One reason for this is that it's a more direct transaction of the storage model (that effects are keyed by their own digest) and the other is that the execution API returns the effects digest:

So actually this is sort of up for debate. Most of the time effects are specifically keyed by the transaction digest that produced them and not the effects digest itself. for example the new KVstore keys effects specifically by TxDigest as well as does the sui-core rocksdb tables. There is one exception to this and that was some interesting interfaces that i built for state-sync, at the time i generally thought that you'd query effects by their own digest and not by the digest of their corresponding txn, that turned out to be inaccurate and this interface is currently unused after some refactoring was done in that area. (@mystenmark if you're reading this you were right :) ) So i'd be inclined to go with your second option and opt to have folks query using the txn digest of their transaction as is done in other blockchains when you query to see if your txn has been included in the chain.

Reply:
I've made the change to the schema here to match what we're talking about here: https://github.com/MystenLabs/sui/pull/13765

I'm perfectly happy with this to be honest -- happy to ship that schema change if others folks are happy.

Reply:
querying a tx by effects digest is not very intuitive to me. In the execution api response we could return the Effects along with many other data, but it's orthogonal to whether users may want to query that by effects' digest. Have we actually seen user requirement about this ability?

Reply:
No there's no user requirement to use effects digest.  The original flow was designed to use effects digest to distinguish between a transaction that was signed/submitted and one that was final, by referring to a data structure that only exists on the network (the effects) by its own digest, but I think this analogy kind of breaks down at the RPC layer, where the RPC is only really aware of the transaction after it's final too.

> In the execution api response we could return the Effects along with many other data, but it's orthogonal to whether users may want to query that by effects' digest.

I'm not too keen to return the effects digest if people can't use it for anything -- that's just asking for people to confuse the two digests, so the PR above entirely removes the effects digest concept from RPC (which matches the situation today, I think).